[說明文件](https://hackmd.io/@shaoeChen/r1CQ9VY98)

載入需求套件

In [54]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import RobustScaler

設定資料

In [55]:
X = np.array([[1, 9], [2, 8], [3, 7], [4, 6], [5, 5], [6, 4], [7, 3], [8, 2], [9, 1]])

In [56]:
X.shape

(9, 2)

上面設置了9筆的假資料，我們使用RobustScaler來看看縮放之後的結果

首先我們以預設置來縮放 

In [57]:
robust = RobustScaler()

In [58]:
X_robust = robust.fit_transform(X)

In [60]:
print("RobustScaler center_: \n" , robust.center_)
print("RobustScaler scale_: \n" , robust.scale_)

RobustScaler center_: 
 [5. 5.]
RobustScaler scale_: 
 [4. 4.]


In [59]:
X_robust

array([[-1.  ,  1.  ],
       [-0.75,  0.75],
       [-0.5 ,  0.5 ],
       [-0.25,  0.25],
       [ 0.  ,  0.  ],
       [ 0.25, -0.25],
       [ 0.5 , -0.5 ],
       [ 0.75, -0.75],
       [ 1.  , -1.  ]])

現在我們來瞭解這個值是怎麼出現的

下表劃出一個表格，我們有九筆資料(第一個row)八個區間(第二個row)，第三個row說明出每個區段的比例，也因此，四分位距中的Q1依著預設值為25%，也就是3，Q3為75%，也就是7，因此IQR，也就是四分位距為7-3=4，而中位數很明顯的就是5

|   	|      	|   	|      	|   	|      	|   	|      	|   	|      	|   	|      	|   	|      	|   	|       	|   	|
|---	|------	|---	|------	|---	|------	|---	|------	|---	|------	|---	|------	|---	|------	|---	|-------	|---	|
| 1 	|      	| 2 	|      	| 3 	|      	| 4 	|      	| 5 	|      	| 6 	|      	| 7 	|      	| 8 	|       	| 9 	|
|   	| 1    	|   	| 2    	|   	| 3    	|   	| 4    	|   	| 5    	|   	| 6    	|   	| 7    	|   	| 8     	|   	|
|   	| 12.5 	|   	| 25.0 	|   	| 37.5 	|   	| 50.0 	|   	| 62.6 	|   	| 75.0 	|   	| 87.5 	|   	| 100.0 	|   	|

因此，上面我們得到的`scale_=4`，試著計算其中一個，以第二個值為範例：
* (2 - 5) / 4 = -3 / 4 = -0.75

第二個特徵可以自行驗證

下面給出一個範例，這是來自[De-yu](https://de-yu-code.blogspot.com/2018/04/sklearnpreprocessingrobustscaler.html)的blogspot

In [61]:
X_1 = np.array([[1 , 5],[2 , 9],[3 , 8],[4 , 3],[5 ,10],[6, 1],[7,6],[8,15],[9,0]])

In [48]:
robust_1 = RobustScaler(quantile_range =(40,75))

In [49]:
X_robust = robust_1.fit_transform(X_1)

In [50]:
X_robust

array([[-1.42857143, -0.26315789],
       [-1.07142857,  0.78947368],
       [-0.71428571,  0.52631579],
       [-0.35714286, -0.78947368],
       [ 0.        ,  1.05263158],
       [ 0.35714286, -1.31578947],
       [ 0.71428571,  0.        ],
       [ 1.07142857,  2.36842105],
       [ 1.42857143, -1.57894737]])

In [62]:
print("RobustScaler center_: \n" , robust_1.center_)
print("RobustScaler scale_: \n" , robust_1.scale_)

RobustScaler center_: 
 [5. 6.]
RobustScaler scale_: 
 [2.8 3.8]


一樣的，劃出表格來做說明

|   	|      	|   	|      	|   	|      	|   	|      	|   	|      	|   	|      	|   	|      	|    	|       	|    	|
|---	|------	|---	|------	|---	|------	|---	|------	|---	|------	|---	|------	|---	|------	|----	|-------	|----	|
| 0 	|      	| 1 	|      	| 3 	|      	| 5 	|      	| 6 	|      	| 8 	|      	| 9 	|      	| 10 	|       	| 15 	|
|   	| 1    	|   	| 2    	|   	| 3    	|   	| 4    	|   	| 5    	|   	| 6    	|   	| 7    	|    	| 8     	|    	|
|   	| 12.5 	|   	| 25.0 	|   	| 37.5 	|   	| 50.0 	|   	| 62.6 	|   	| 75.0 	|   	| 87.5 	|    	| 100.0 	|    	|

中位數很明顯的就是6，而四分距的Q3就是9，其中Q1因為我們在初始化的時候調整為40%，因此40%是介於5~6之間，其中於5的距離為40 - 37.5 = 2.5，於6的距離為50 - 40 = 10，5的部份比較近也因此有較大的權重，因此：
* Q3 = 6 x 2.5 / 12.5 + 5 x 10 / 12.5 = 1.2 + 4 = 5.2

四分位距 = Q3 - Q1 = 9 - 5.2 = 3.8，與上面程式執行所得相同，另一維度的部份也可以用相同的作法來驗證，這部份就留給使用者自己來動手計算看看了。